### ISSUE TITLE
Intermittent "Client is offline" rejection from Realtime Database `get()`

### ISSUE THREAD

**Issue by XuluWarrior** [firebase/firebase-js-sdk#5612]
### [REQUIRED] Describe your environment

  * Operating System version: AWS Lambda - node-14
  * Firebase SDK version: 9.1.1
  * Firebase Product: database

### [REQUIRED] Describe the problem
Since updating my code to use Firebase 9.x and the new modular API, I am getting intermittent "Client is offline" promise rejections when calling `get()`.
There is a Firebase [Google Group](https://groups.google.com/g/firebase-talk/c/VpQms_TnBOw?pli=1) that has other users that have experienced the same problem but I can't find a related issue in GitHub, so I thought I would raise one.

I currently running 9.1.1 and so will try upgrading to 9.1.2.  I will also enable logging.
#### Relevant Code:

```javascript
initializeApp(config)
database = await getDatabase();
selectionsRef = ref(database, path);
await get(child(selectionsRef, systemName), "value");
```

Call stack:
```
Error: Error: Client is offline.
  File "/var/task/node_modules/@firebase/database/dist/index.node.cjs.js", line 11517, col 31, in null.<anonymous>
    return Promise.reject(new Error(err));
```

**Comment by jbalidiong** [comment]
Hi @XuluWarrior, thanks for reaching out. I tried replicating, but I wasn't able to reproduce the error. If I can replicate the issue, I can have a better look into it. Please share a minimal, but complete sample of a project that I can run locally.

**Comment by schmidt-sebastian** [comment]
@IanWyszynski Can you take a look?

**Comment by qlereboursBS** [comment]
I have the same issue.
I was not able to reproduce it on my local machine, but it's reproducible on a distant environment:
 - I'm authenticating to the website, which redirects me to a page loading my user data
 - The error doesn't show when loading data
 - I can refresh as many times as I want, the error won't show
 - If I wait a few minutes or a few hours, when refreshing the page, the error is showing

![image](https://user-images.githubusercontent.com/27200095/141645572-670fc2b7-59f6-4330-93a4-d2a99575312a.png)

You can reproduce on the dev environment of my website: dev.traveledmap.com
Since I was able to reproduce it only when logged for the moment, I think you would have to create an account.

**Comment by qlereboursBS** [comment]
I was just able to reproduce it on my local machine after waiting for a few minutes to refresh the page. However I had the error just once, and then the data successfully loaded (which was not the case on the screen I sent)

**Comment by qlereboursBS** [comment]
I was able to have kind of a reproduction on a new project, but I think that it's the expected behavior:
 - I'm putting a breakpoint in the `ready_` method
 - I refresh the page that will load some data (`get(ref(getDatabase(), 'my-ref'))`
 - I'm waiting for a few seconds, and the release the breakpoint
 - I can see the error in the console.
I'm not sure if it's the expected behavior because the database request is done just after I release the breakpoint and however the request fails

Maybe it could be related to our issue?

https://user-images.githubusercontent.com/27200095/141649559-d778bd1b-27f4-4ba9-921c-e0c8a859e77c.mp4

**Comment by qlereboursBS** [comment]
I was able to have the issue with my simple example, without breakpoints:
 - clone this repository: https://github.com/qlereboursBS/firebase-client-offline-reproduction
 - replace the two "TODOs" in the code with your firebase config, and the path to the data you want to fetch
 - run `yarn && yarn dev` to start the project
 - Hard refresh (ctrl + maj + R on Chrome) the page approximately 10 times, the error should appear randomly (I faced it approximately every 6 times)

@jbalidiong @schmidt-sebastian can you have a look please?
To my mind, this is a high priority bug since it should impact all projects using realtime database

**Comment by jbalidiong** [comment]
Hi @qlereboursBS, thanks for adding information regarding this issue. I just want to confirm if the repository you've provided contains any code with Firebase? Currently, I'm seeing the default files when created with Next.js. Maybe some files are not uploaded in the repository.

**Comment by qlereboursBS** [comment]
I'm really sorry, I just realized that the commit and push didn't work, I just pushed it.
You will find the code in `pages/_app.js`

**Comment by qlereboursBS** [comment]
Were you able to reproduce @jbalidiong ?
It's a blocking issue for me since it's already on production due to the migration to the v9 and I would like to prevent rolling back before the migration

**Comment by jbalidiong** [comment]
@qlereboursBS, unfortunately, I wasn't able to replicate it. I also waited for a few minutes before refreshing the page and I didn't encounter the error. Let me check with other engineers and see if they can replicate it on their end.

**Comment by qlereboursBS** [comment]
I think it may be related to a poor connection. I'm currently in a hotel, the connection is poor, I can access pages but it's very frequent to have this error (almost everytime). When I created the issue, I was at home with a better connection (not the best, but good enough).
What is wierd is that I can have the error for a first request, and get data from another database request just after

**Comment by harveylee** [comment]
I was having a problem with similar symptoms, using firebase-admin on Windows WSL.

After I enabled logging on Firebase, it turned out that the problem under the hood was that my auth token was being rejected (and fwiw, due to an issue with clock being out of sync: https://github.com/microsoft/WSL/issues/4245).

So I wonder a couple of things:
- Is @qlereboursBS's poor connection leading to a clock/timing issue on auth token refreshes?
- If the issue is due to a rejected auth token (as it was in my case), a more specific error message in the first place would have helped diagnose the problem

**Comment by danielo515** [comment]
The way I have to reproduce it is to put a breakpoint and just take your time to analyze what is happening.
Since get is supposed to be asynchronous just run it and put the breakpoint in any other area of the app. Then this error occurs.

**Comment by JoshuaTyler-GitHub** [comment]
I have been able to reliably recreate this issue every time using a basic Firebase Database GET with the Realtime Database.

`      FirebaseDatabase.getDatabase()
      .then((database) => {
        const dbRef = ref(database);
        get(child(dbRef, path))
        .then((snapshot) => {
          if(snapshot.exists()) resolve(snapshot.val());
          else reject(ERROR_UNDEFINED_VALUE);
        })
        .catch((error) => reject(error));
      });`
      
In Chrome developer tools, throttle your connection speed to mid or low tier and it will throw "Error: Client is offline."

**Comment by navinssh** [comment]
I had similar issue, eventually I figured I had setup the realtime database after I had already generated the Firebase config so it was missing the `databaseURL` from the config.

Find the Firebase config again from the app it should have it in the config. 

Hope this helps.

**Comment by Javaworld22** [comment]
The issue is the database allows one person at a time. Hence, if a method is reading from a database, this method locks the database from another method reading from the same application

**Comment by nugmanoff** [comment]
> I have been able to reliably recreate this issue every time using a basic Firebase Database GET with the Realtime Database.
> 
> ` FirebaseDatabase.getDatabase() .then((database) => { const dbRef = ref(database); get(child(dbRef, path)) .then((snapshot) => { if(snapshot.exists()) resolve(snapshot.val()); else reject(ERROR_UNDEFINED_VALUE); }) .catch((error) => reject(error)); });`
> 
> In Chrome developer tools, throttle your connection speed to mid or low tier and it will throw "Error: Client is offline."

I have the same issue, ping me if you figure out something

**Comment by ittaboba** [comment]
Almost the same for me on v9.0.1 as mentioned by @qlereboursBS 

> I think it may be related to a poor connection. I'm currently in a hotel, the connection is poor, I can access pages but it's very frequent to have this error (almost every time). When I created the issue, I was at home with a better connection (not the best, but good enough). What is weird is that I can have the error for a first request, and get data from another database request just after

My home wifi is currently at ~5 Mbps in download (don't know why 🤷‍♂️) and I get the error. If I switch to hotspot at ~40 Mbps it works.

This is very annoying and unpredictable behavior. Could you please fix it?

**Comment by ittaboba** [comment]
In the meantime, I have found a workaround by replacing get calls with this. Hope it helps.

```
function getData(ref: DatabaseReference): Promise<DataSnapshot> {
    return new Promise((resolve, reject) => {
        function onError(error) {
            reject(error);
        } 

        function onData(snap) {
            resolve(snap);
        }
  
        onValue(ref, onData, onError, { onlyOnce: true });
    });
}
```

**Comment by signorbusi** [comment]
We're facing the same problem in our application and it's definitely related to unstable / poor internet connection. This is a major problem for us and very unpredictable. It leads to random crashes of our application.

**Comment by rohanx96** [comment]
> I have been able to reliably recreate this issue every time using a basic Firebase Database GET with the Realtime Database.
> 
> ` FirebaseDatabase.getDatabase() .then((database) => { const dbRef = ref(database); get(child(dbRef, path)) .then((snapshot) => { if(snapshot.exists()) resolve(snapshot.val()); else reject(ERROR_UNDEFINED_VALUE); }) .catch((error) => reject(error)); });`
> 
> In Chrome developer tools, throttle your connection speed to mid or low tier and it will throw "Error: Client is offline."

Able to consistently reproduce the issue this way for our application too. We have a Firebase fetch when our webpage loads. Throttling network to slow causes the Firebase fetch to fail with error Client is offline. This also happens if multiple (20-25) tabs of our application are opened simultaneously, most likely causing the network bandwidth to be divided and resulting in a slow throttling like scenario.

**Comment by maneesht** [comment]
This should be fixed in the latest version of the SDK with #6340.

**Comment by Ashwith2001** [comment]
mmm
